tests/Aspire.Hosting.Tests/ApplicationModel/ReferenceExpressionTests.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+
+using Xunit;
+
+namespace Aspire.Hosting.Tests.ApplicationModel;
+public class ReferenceExpressionTests
+{
+    [Theory]
+    [InlineData("{x}", "{x}")]
+    [InlineData("{{x}}", "{x}")]
+    [InlineData("{x", "{x")]
+    [InlineData("x}", "x}")]
+    [InlineData("{1 var}", "{1 var}")]
+    [InlineData("{var 1}", "{var 1}")]
+    [InlineData("{1myVar}", "{1myVar}")]
+    [InlineData("{myVar1}", "{myVar1}")]
+    public void ReferenceExpressionHandlesValueWithNonParameterBrackets(string input, string expected)
+    {
+        var expr = ReferenceExpression.Create($"{input}").ValueExpression;
+        Assert.Equal(expected, expr);
+    }
+
+    [Theory]
+    [InlineData("{0}", new string[] { "abc123" }, "abc123")]
+    [InlineData("{0} test", new string[] { "abc123" }, "abc123 test")]
+    [InlineData("test {0}", new string[] { "abc123" }, "test abc123")]
+    [InlineData("https://{0}:{1}/{2}?key={3}", new string[] { "test.com", "443", "path", "1234" }, "https://test.com:443/path?key=1234")]
+    public void ReferenceExpressionHandlesValueWithParameterBrackets(string input, string[] parameters, string expected)
+    {
+        var expr = ReferenceExpression.Create($"{input}", [new HostUrl("test")], parameters).ValueExpression;
+        Assert.Equal(expected, expr);
+    }
+
+    public static readonly object[][] ValidFormattingInParameterBracketCases = [
+        ["{0:D}", new DateTime(2024,05,22), string.Format(CultureInfo.InvariantCulture, "{0:D}", new DateTime(2024, 05, 22).ToString())],
+        ["{0:N}", 123456.78, string.Format(CultureInfo.InvariantCulture, "{0:N}", "123456.78")]
+    ];
+
+    [Theory, MemberData(nameof(ValidFormattingInParameterBracketCases))]
+    public void ReferenceExpressionHandlesValueWithFormattingInParameterBrackets(string input, string parameterValue, string expected)
+    {
+        var expr = ReferenceExpression.Create($"{input}", [new HostUrl("test")], [parameterValue]).ValueExpression;
+        Assert.Equal(expected, expr);
+    }
+
+    [Theory]
+    [InlineData("{0} {x}", "abc123", "abc123 {x}")]
+    [InlineData("{x} {0}", "abc123", "{x} abc123")]
+    [InlineData("{0} test {x}", "abc123", "abc123 test {x}")]
+    [InlineData("{x} test {0}", "abc123", "{x} test abc123")]
+    public void ReferenceExpressionHandlesValueWithBothParameterAndNonParameterBrackets(string input, string parameterValue, string expected)
```

Comment:
Based on my understanding, these 4 test cases are not valid and should actually fail. Unlike with the other `Create` override, this one does not treat the input string as a `ExpressionInterpolatedStringHandler`. It's just a string, and is expected to be a valid format string. The presence of the `{x}` in the format string makes it blow up with "Input string was not in a correct format. Failure to parse near offset 5. Expected an ASCII digit.", which I think is fully expected.

Reply:
My understanding of #3787 was to be able to handle when the string had formatting like this and handle / accept it without throwing the exception.

Reply:
Yes, but I think that only applies to the `ReferenceExpression.Create($"{s}")`, which uses `ExpressionInterpolatedStringHandler`, and not for this overload, which takes a plain format string. In this case, there is no difference between passing `$"{input}"` as you do, and just passing `input`.

That being said, I very well might be wrong, hence waiting for feedback from Aspire folks like @mitchdenny and @davidfowl.